examples/blockchain/network.rs

```diff
@@ -0,0 +1,364 @@
+use crate::chain::Block;
+use aleph_bft::NodeIndex;
+use codec::{Decode, Encode};
+use futures::{
+    channel::{
+        mpsc::{self, UnboundedReceiver, UnboundedSender},
+        oneshot,
+    },
+    prelude::*,
+    Future, FutureExt, StreamExt,
+};
+
+use log::{debug, info};
+
+use std::{collections::HashMap, error::Error, io, iter, time::Duration};
+
+use libp2p::{
+    development_transport, identity,
+    mdns::{Mdns, MdnsConfig, MdnsEvent},
+    request_response::{
+        ProtocolSupport, RequestResponse, RequestResponseCodec, RequestResponseConfig,
+        RequestResponseEvent, RequestResponseMessage,
+    },
+    swarm::{NetworkBehaviourEventProcess, SwarmBuilder},
+    NetworkBehaviour, PeerId, Swarm,
+};
+
+use crate::{
+    chain::Data,
+    crypto::{Hasher256, PartialMultisignature, Signature},
+};
+
+#[derive(Clone)]
+pub(crate) struct Spawner;
+
+impl aleph_bft::SpawnHandle for Spawner {
+    fn spawn(&self, _: &str, task: impl Future<Output = ()> + Send + 'static) {
+        tokio::spawn(task);
+    }
+}
+
+const ALEPH_PROTOCOL_NAME: &str = "/alephbft/test/1";
+
+type NetworkData = aleph_bft::NetworkData<Hasher256, Data, Signature, PartialMultisignature>;
+
+#[allow(clippy::large_enum_variant)]
+#[derive(Clone, Encode, Decode)]
+enum Message {
+    Auth(NodeIndex),
+    Consensus(NetworkData),
+    Block(Block),
+}
+
+enum MessageRecipient {
```

Comment:
Eh, it's a bit annoying – we already have this type inside AlephBFT (with slightly different names), and maybe just exposing it would be simpler. But it doesn't really make sense to expose it, because of the `Network` interface having `broadcast` as a separate method, rather than using `send(msg, Recipient)`. Maybe we should think about refactoring it into the latter shape in the future?

For context the aleph-node network implementation also uses this abstraction, although with the approach you had here before, with `None` meaning everyone.

Reply:
Yeah, a little annoying, although these are internals of some example we will probably run once or a few times and forget about it. Whether we should refactor the Network trait: I like both versions actually, the one you propose probably allows the trait implementations to be a few lines shorter :)

Reply:
I will add a task for that, maybe one day~.